packages/uix-core/src/tunnel/tunnel-messenger.ts

```diff
@@ -0,0 +1,99 @@
+import { NS_ROOT, VERSION } from "../constants";
+import { isPlainObject } from "../value-assertions";
+import { WrappedMessage, isWrapped, wrap, unwrap } from "../message-wrapper";
+import { HandshakeAcceptedTicket, HandshakeOfferedTicket } from "../tickets";
+
+type Handshake = HandshakeAcceptedTicket | HandshakeOfferedTicket;
+type HandshakeAccepted = WrappedMessage<HandshakeAcceptedTicket>;
+type HandshakeOffered = WrappedMessage<HandshakeOfferedTicket>;
+type HandshakeMessage = HandshakeAccepted | HandshakeOffered;
+
+export class TunnelMessenger {
+  private myOrigin: string;
+  private remoteOrigin: string;
+  private logger: Console;
+  private versionWarnings = new Set<string>();
+  constructor(opts: {
+    myOrigin: string;
+    targetOrigin: string;
+    logger: Console;
+  }) {
+    this.myOrigin = opts.myOrigin;
+    this.remoteOrigin =
+      opts.targetOrigin === "*" ? "remote document" : opts.targetOrigin;
+    this.logger = opts.logger;
+  }
+  resetWarnings() {
+    this.versionWarnings.clear();
+  }
+
+  makeAccepted(id: string): HandshakeAccepted {
+    return wrap({
+      accepts: id,
+      version: VERSION,
+    });
+  }
+  makeOffered(id: string): HandshakeOffered {
+    return wrap({
+      offers: id,
+      version: VERSION,
+    });
+  }
+  isHandshakeAccepting(
+    message: unknown,
+    id: string
+  ): message is HandshakeAccepted {
+    return (
+      this.isHandshake(message) &&
+      unwrap(message as HandshakeAccepted).accepts === id
+    );
+  }
+  isHandshakeOffer(message: unknown): message is HandshakeOffered {
+    return (
+      this.isHandshake(message) &&
+      typeof unwrap(message as HandshakeOffered).offers === "string"
+    );
+  }
+  isHandshake(message: unknown): message is HandshakeMessage {
+    if (!isWrapped(message)) {
+      this.logMalformed(message);
+      return false;
+    }
+    const tunnelData: Handshake = unwrap<Handshake>(
+      message as HandshakeMessage
+    );
+    if (
+      !isPlainObject(tunnelData) ||
+      typeof tunnelData.version !== "string" ||
+      !(Reflect.has(tunnelData, "accepts") || Reflect.has(tunnelData, "offers"))
+    ) {
+      this.logMalformed(message);
+      return false;
+    }
+    const { version } = tunnelData;
+    if (version !== VERSION && !this.versionWarnings.has(version)) {
```

Comment:
Should we always expect exact versions match or we can allow patch or minor version differences?

Reply:
(following in-person discussion) The "safe" version of this feature would require adding the `semver` library, which is 88K. Maybe we could assume that production version strings will always be `/\d+\.\d+\.\d+/`.

However! While we're still in `v0.x.x`, we have no API guarantees, and we can defer that question and just warn for any difference for now.